src/transport/message_handler.rs

```diff
@@ -0,0 +1,270 @@
+use crate::transport::{jsonrpc, msgs};
+use bitcoin::secp256k1::PublicKey;
+use lightning::{
+	ln::{peer_handler::CustomMessageHandler, wire::CustomMessageReader},
+	log_info,
+	util::logger::Logger,
+};
+use std::{collections::HashMap, io, ops::Deref, sync::Mutex};
+
+pub trait LSPMessageHandler {
+	fn handle_lsps_request(
+		&self, request: jsonrpc::Request, counterparty_node_id: &PublicKey,
+	) -> Result<(), lightning::ln::msgs::LightningError>;
+	fn handle_lsps_response(
+		&self, response: jsonrpc::Response, counterparty_node_id: &PublicKey,
+	) -> Result<(), lightning::ln::msgs::LightningError>;
+	fn handle_lsps_notification(
+		&self, notification: jsonrpc::Notification, counterparty_node_id: &PublicKey,
+	) -> Result<(), lightning::ln::msgs::LightningError>;
+	fn get_and_clear_pending_msg(&self) -> Vec<(PublicKey, msgs::RawLSPSMessage)>;
+	fn get_protocol_number(&self) -> u16;
+}
+
+pub struct LSPManager<LMH: Deref, L: Deref>
+where
+	LMH::Target: LSPMessageHandler,
+	L::Target: Logger,
+{
+	pending_messages: Mutex<Vec<(PublicKey, msgs::RawLSPSMessage)>>,
+	pending_requests: Mutex<HashMap<String, jsonrpc::Request>>,
+	message_handlers: Mutex<HashMap<String, LMH>>,
+	logger: L,
+}
+
+impl<LMH: Deref, L: Deref> LSPManager<LMH, L>
+where
+	LMH::Target: LSPMessageHandler,
+	L::Target: Logger,
+{
+	pub fn new(logger: L) -> Self {
+		Self {
+			pending_messages: Mutex::new(Vec::new()),
+			pending_requests: Mutex::new(HashMap::new()),
+			message_handlers: Mutex::new(HashMap::new()),
+			logger,
+		}
+	}
+
+	pub fn register_message_handler(&self, prefix: String, message_handler: LMH) {
+		self.message_handlers.lock().unwrap().insert(prefix, message_handler);
+	}
+
+	pub fn list_protocols(
+		&self, counterparty_node_id: PublicKey,
+	) -> Result<(), lightning::ln::msgs::LightningError> {
+		let request = jsonrpc::Request {
+			jsonrpc: jsonrpc::VERSION.to_string(),
+			id: serde_json::json!(1),
```

Comment:
TODO: Clients MUST use a high-entropy id string for JSON-RPC 2.0 requests, such as a UUID, or a hex encoding of a random binary blob of at least 80 bits, with randomness acquired from a cryptographically secure source. Clients MUST NOT use just a simple incrementing counter for the id.

Reply:
Made handler generic over EntropySource to get random bytes and convert to hex string to use as id.  For now taking first 16 bytes of the 32 byte blob as spec says 80 bits is enough and I think standard UUID is 16 bytes.  Probably still overkill but 🤷

Reply:
Not overkill: none of the implementations actually allow a non-racy way to determine if a custom message is sent before or after a disconnection.

Consider that a disconnection might imply the client crashed and restarted.  If it used an in-memory incrementing counter, then the counter might get reset, meaning the client will duplicate IDs, and there is the possibility that the LSP *then* sends a message (since none of the send-custom-message APIs in any of LDK, LDN, and CLN allow non-racy way to ensure sending on the same connection a previous custom message was received on, everybody just keys on node ID) that is for the previous connection session (the one that crashed) which might be confused as the response to a newly-sent request.

Your alternative is a persisted counter (which requires you to actually do `flush`ed disk I/O every time the client makes a request) or a randomly-picked nonce (which just requires a lock on your entropy source and some CPU cycles to update the PRNG).  80 bits is absolute minimum necessary for randomly-picked nonce IMO.  Could get away with 64 bits but I like me nice cheap margins.